### PR TITLE
[ADD] VNode, Format: basically handle many nodes

### PR DESCRIPTION
**Pull request by Zinston**
Many of these will not change. Some will change significantly (`table`). These are placeholders for many of the most common nodes and formats from the DOM:
- Format: Link (a, button)
- Divider (div)
- Image (img)
- Format: Subscript (sub)
- Format: Superscript (sup)
- Blockquote (blockquote)
- Youtube (iframe src="...youtu...")
- Table (table) - naive: no contents
- Metadata (script, style, ...) - naive: no rendering

This also includes plenty of fixes of problems that were revealed in the process of implementing these.

### CODE REVIEW

File: examples/jabberwockipedia/index.html [hunk 1]
```diff
@@ -12,24 +12,31 @@ <h1 id="firstHeading" class="firstHeading" lang="en">Jabberwocky</h1>
                     accepted revision</a>, <a class="external text"
                     href="https://en.wikipedia.org/w/index.php?title=Special:Log&amp;type=review&amp;page=Jabberwocky">reviewed</a>
                 on <i>26 January 2020</i>.</div>
-            <a class="mw-jump-link" href="#mw-head">Jump to navigation</a>
-            <a class="mw-jump-link" href="#p-search">Jump to search</a>
+            <p>
+                <a class="mw-jump-link" href="#mw-head">Jump to navigation</a>
+                <a class="mw-jump-link" href="#p-search">Jump to search</a>
+            </p>
```
Comment (dmo-odoo) [inline examples/jabberwockipedia/index.html]: What is the problem that this commit is avoiding ?

File: packages/plugin-technical/TechnicalNode.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { MarkerNode } from '../core/src/VNodes/MarkerNode';
+
+export class TechnicalNode extends MarkerNode {
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-technical/TechnicalNode.ts]: image

File: packages/utils/src/utils.ts [hunk 3]
```diff
@@ -13,7 +13,7 @@ export function isBlock(node: Node): boolean {
     document.body.appendChild(temporaryElement);
     const display = window.getComputedStyle(temporaryElement).display;
     document.body.removeChild(temporaryElement);
-    return display.includes('block') || display.includes('list');
+    return !display.includes('inline');
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/utils/src/utils.ts]: Missing 'contents'

File: packages/core/src/VRange.ts [hunk 4]
```diff
@@ -253,7 +253,17 @@ export class VRange {
...
             }
-            this.endContainer.mergeWith(this.startContainer);
+            if (this.endContainer.ancestors().includes(this.startContainer)) {
+                const endContainer = this.endContainer;
+                for (const child of endContainer.children.slice()) {
+                    endContainer.before(child);
+                }
+                endContainer.remove();
+            } else if (this.startContainer.ancestors().includes(this.endContainer)) {
+                // TODO
```
Comment (Zinston) [inline packages/core/src/VRange.ts]: @dmo-odoo Here we go!

File: packages/plugin-devtools/src/components/TreeComponent.ts [hunk 5]
```diff
@@ -142,6 +142,14 @@ export class TreeComponent extends OwlUIComponent<NodeProps> {
     isUnderline(node: VNode): boolean {
         return node.is(InlineNode) && !!node.formats.find(format => format.name === 'u');
     }
+    /**
+     * Return true if the given node is underline.
+     *
+     * @param node
+     */
+    isLink(node: VNode): boolean {
+        return node.is(InlineNode) && !!node.formats.find(format => format.name === 'link');
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-devtools/src/components/TreeComponent.ts]: Handled later.

File: packages/plugin-inline/Formats.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    /**
+     * Return the first format in the array that is an instance of the given
+     * FormatClass, if any.
+     *
+     * @param format
+     */
+    get<T extends Format>(format: T | Constructor<T>): T {
+        if (format instanceof Format) {
+            return this.find(formatInstance => formatInstance === format) as T;
```
Comment (Zinston) [inline packages/plugin-inline/Formats.ts]: Well this is ridiculous. It will always return itself!

File: packages/plugin-inline/Formats.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        } else {
+            return this.find(formatInstance => formatInstance instanceof format) as T;
+        }
+    }
+    /**
+     * Return the index of the first format in the array that is an instance of
```
Comment (Zinston) [inline packages/plugin-inline/Formats.ts]: Comments need to be updated since it can take an instance as well.

File: packages/plugin-char/Char.ts [hunk 8]
```diff
@@ -79,17 +79,12 @@ export class Char extends JWPlugin {
         if (!this.formatCache) {
             this.formatCache = this.getCurrentFormats();
         }
-        const index = this.formatCache.findIndex(f => f instanceof FormatClass);
-        if (index !== -1) {
-            this.formatCache.splice(index, 1);
-        } else {
-            this.formatCache.push(new FormatClass());
-        }
+        this.formatCache.toggle(FormatClass);
```
Comment (Zinston) [inline packages/plugin-char/Char.ts]: BIM! :boom:

File: packages/plugin-char/Char.ts [hunk 9]
```diff
@@ -101,10 +96,10 @@ export class Char extends JWPlugin {
             inlineToCopyFormat = range.start.nextSibling();
         }
         if (inlineToCopyFormat && inlineToCopyFormat.is(InlineNode)) {
-            return [...inlineToCopyFormat.formats];
+            return inlineToCopyFormat.formats.clone();
```
Comment (Zinston) [inline packages/plugin-char/Char.ts]: BOOM! :bomb:

File: packages/plugin-youtube/YoutubeDomParser.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    predicate = (item: Node): boolean => {
+        return (
+            item instanceof Element &&
+            item.nodeName === 'IFRAME' &&
+            item.getAttribute('src').includes('youtu')
+        );
+    };
+
+    async parse(item: Element): Promise<YoutubeNode[]> {
+        const image = new YoutubeNode();
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-youtube/YoutubeDomParser.ts]: image

**Comment by Zinston** [comment]
@dmo-odoo Ready and rebased.

**Comment by dmo-odoo** [comment]
🎉

**Comment by Zinston** [comment]
🍾

**Pull request merged**